clang/lib/Sema/SemaExpr.cpp

```diff
@@ -14201,6 +14199,15 @@ QualType Sema::CheckAddressOfOperand(ExprResult &OrigOp, SourceLocation OpLoc) {
 
   CheckAddressOfPackedMember(op);
 
+  if (getLangOpts().C99) {
```

Comment:
I'm curious why this code is not placed where the comment was removed.  Why place it here?

Is this a change that is specific to the Checked C version of clang?  Or should it be propagated upstream?  Put another way, is this a bug in clang?  Or is there something specific about Checked C that requires this change?

Reply:
This code was placed here (rather than where the comment was removed) so that it occurs after the check on [line 14127](https://github.com/microsoft/checkedc-clang/blob/master/clang/lib/Sema/SemaExpr.cpp#L14127) for taking the address of an lvalue. If this code is placed where the comment is removed, then certain tests (e.g. Sema/complex-imag.c) fail due to missing expected errors (they expect errors "cannot take the address of an rvalue of type <T>" to be emitted).

I can add a comment explaining why this code is placed where it is.

To the best of my knowledge, this change isn't Checked C specific.

Reply:
If this code is moved to where the comment was removed, then the Sema/expr-address-of.c test fails. This test expects the following error:

```
void foo() {
  register int x[10];
  &x[10];              // expected-error {{address of register variable requested}}
}
```

The "address of register variable requested" error is emitted in the call to `diagnoseAddressOfInvalidType` on [line 14187](https://github.com/microsoft/checkedc-clang/blob/master/clang/lib/Sema/SemaExpr.cpp#L14187). If the code in this PR is added before line 14187, then this expected error will not be emitted.

Reply:
I looked at the code you are adding and I don't think it is needed for the regular C path.   Taking the address of a subscript expression `e1[e2]`, where e1 or e2 is a pointer to T, will result in a pointer to T being created with the existing code.  The subscript expression will have type T and the final statement will create a pointer to T.    

I believe the code is really only needed for the Checked C path.  It might be better to put this under a Checked C flag.   You could then explain in the comment that the code avoids the unexpected result of &e1[e2] having a different kind of pointer type than the pointer type that is being subscripted.  This can happen in unchecked scopes where the `&` operator is applied to a subscript expression involving a checked pointer.